File: app/src/main/java/activity/SeguindoActivity.java
package activity;

import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.crash.FirebaseCrash;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

import adapter.SeguindoAdapater;
import br.com.icaro.filme.R;
import domain.UserTvshow;
import info.movito.themoviedbapi.TmdbApi;
import info.movito.themoviedbapi.TmdbTV;
import info.movito.themoviedbapi.TmdbTvSeasons;
import info.movito.themoviedbapi.model.tv.TvSeason;
import info.movito.themoviedbapi.model.tv.TvSeries;
import utils.Config;
import utils.UtilsApp;

import static utils.UtilsApp.setEp;
import static utils.UtilsApp.setUserTvShow;

/**
 * Created by icaro on 25/11/16.
 */
public class SeguindoActivity extends BaseActivity {

    private final String TAG = SeguindoActivity.class.getName();
    private ViewPager viewPager;
    private TabLayout tabLayout;
    private ProgressBar progressBar;
    private LinearLayout linearLayout;
    private DatabaseReference seguindoDataBase;
    private ValueEventListener eventListener;
    private List<UserTvshow> userTvshows;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_usuario_list);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        setUpToolBar();
        getSupportActionBar().setTitle(R.string.seguindo);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        viewPager = (ViewPager) findViewById(R.id.viewpage_usuario);
        tabLayout = (TabLayout) findViewById(R.id.tabLayout);
        progressBar = (ProgressBar) findViewById(R.id.progress);
        linearLayout = (LinearLayout) findViewById(R.id.linear_usuario_list);

//        AdView adview = (AdView) findViewById(R.id.adView);
//        AdRequest adRequest = new AdRequest.Builder()
//                .addTestDevice(AdRequest.DEVICE_ID_EMULATOR)        // All emulators
//                .addTestDevice("AC98C820A50B4AD8A2106EDE96FB87D4")  // An example device ID
//                .build();
//        adview.loadAd(adRequest);

        if (UtilsApp.isNetWorkAvailable(this)) {

            iniciarFirebases();
            setEventListenerSeguindo();
        } else {
            snack();
        }
    }

    private void iniciarFirebases() {

        FirebaseAuth mAuth = FirebaseAuth.getInstance();
        FirebaseDatabase database = FirebaseDatabase.getInstance();

        seguindoDataBase = database.getReference("users").child(mAuth.getCurrentUser()
                .getUid()).child("seguindo");
    }

    protected void snack() {
        Snackbar.make(linearLayout, R.string.no_internet, Snackbar.LENGTH_INDEFINITE)
                .setAction(R.string.retry, new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if (UtilsApp.isNetWorkAvailable(getBaseContext())) {
                            //text_elenco_no_internet.setVisibility(View.GONE);
                            iniciarFirebases();
                            setEventListenerSeguindo();
                        } else {
                            snack();
                        }
                    }
                }).show();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }


    @SuppressWarnings("deprecation")
    private void setupViewPagerTabs() {

        viewPager.setOffscreenPageLimit(1);
        viewPager.setCurrentItem(0);
        tabLayout.setupWithViewPager(viewPager);
        tabLayout.setSelectedTabIndicatorColor(getResources().getColor(R.color.accent));
        viewPager.setAdapter(new SeguindoAdapater(SeguindoActivity.this, getSupportFragmentManager(),
                userTvshows));
    }

    private void setEventListenerSeguindo() {
        eventListener = new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                userTvshows = new ArrayList<>();
                if (dataSnapshot.exists()) {

                    for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                        UserTvshow userTvshow = snapshot.getValue(UserTvshow.class);
                        atualizarRealDate(userTvshow);
                        userTvshows.add(userTvshow);
                    }
                    setupViewPagerTabs();
                }
                progressBar.setVisibility(View.GONE);
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        };

        seguindoDataBase.addListenerForSingleValueEvent(eventListener);
        //Chamando apenas uma vez, necessario? não poderia deixar o firebases atualizar?
    }

    private void atualizarRealDate(final UserTvshow userTvshowOld) {
        try {
            new Thread(new Runnable() {
                @Override
                public void run() {
                    TmdbTvSeasons tvSeasons = new TmdbApi(Config.TMDB_API_KEY).getTvSeasons();
                    TvSeries series = new TmdbApi(Config.TMDB_API_KEY).getTvSeries()
                            .getSeries(userTvshowOld.getId(), "en", TmdbTV.TvMethod.external_ids);
                    if (userTvshowOld.getNumberOfEpisodes() < series.getNumberOfEpisodes()) {
                        UserTvshow userTvshow = setUserTvShow(series);

                        for (int i = 0; i < series.getSeasons().size(); i++) {
                            TvSeason tvS = series.getSeasons().get(i);
                            TvSeason tvSeason = tvSeasons.getSeason(series.getId(), tvS.getSeasonNumber(), "en", TmdbTvSeasons.SeasonMethod.external_ids); //?
                            userTvshow.getSeasons().get(i).setUserEps(setEp(tvSeason));
                            // Atualiza os eps em userTvShow
                        }
                        if (userTvshowOld.getSeasons() != null && userTvshow.getSeasons() != null) {
                            for (int i = 0; i < userTvshowOld.getSeasons().size(); i++) {
                                userTvshow.getSeasons().get(i).setId(userTvshowOld.getSeasons().get(i).getId());
                                userTvshow.getSeasons().get(i).setSeasonNumber(userTvshowOld.getSeasons().get(i).getSeasonNumber());
                                userTvshow.getSeasons().get(i).setVisto(userTvshowOld.getSeasons().get(i).isVisto());
                                //Atualiza somente os campos do temporada em userTvShow
                            }
                        }

                        for (int i = 0; i < userTvshowOld.getNumberOfSeasons(); i++) {
                            //Log.d(TAG, "Numero de eps - " + userTvshow.getSeasons().get(i).getUserEps().size());
                            if (userTvshow.getSeasons().get(i).getUserEps() != null && userTvshowOld.getSeasons().get(i).getUserEps() != null) {
                                if (userTvshow.getSeasons().get(i).getUserEps().size() > userTvshowOld.getSeasons().get(i).getUserEps().size())
                                    userTvshow.getSeasons().get(i).setVisto(false);
                                //  Se huver novos ep. coloca temporada com não 'vista'
                            }
                            if (userTvshowOld.getSeasons().get(i).getUserEps() != null)
                                for (int i1 = 0; i1 < userTvshowOld.getSeasons().get(i).getUserEps().size(); i1++) {
                                    if (i1 < userTvshowOld.getSeasons().get(i)
                                            .getUserEps().size() && i1 < userTvshow.getSeasons().get(i).getUserEps().size() )
                                        if (userTvshow.getSeasons().get(i) != null && userTvshowOld.getSeasons().get(i) != null)
                                        userTvshow.getSeasons().get(i).getUserEps().set(i1, userTvshowOld.getSeasons().get(i).getUserEps().get(i1));

                                    //coloca as informações antigas na nova versão dos dados.
                                }
                        }
                        seguindoDataBase
                                .child(String.valueOf(series.getId()))
                                .setValue(userTvshow)
                                .addOnCompleteListener(new OnCompleteListener<Void>() {
                                    @Override
                                    public void onComplete(@NonNull Task<Void> task) {
                                        if (task.isSuccessful()) {
                                            runOnUiThread(new Runnable() {
                                                @Override
                                                public void run() {
                                                    Toast.makeText(SeguindoActivity.this, R.string.season_updated, Toast.LENGTH_SHORT).show();
                                                }
                                            });
//
                                        }
                                    }
                                });
                    }
                }
            }).start();
        } catch (Exception e) {
            FirebaseCrash.report(e);
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    Toast.makeText(SeguindoActivity.this, R.string.ops_seguir_novamente, Toast.LENGTH_SHORT).show();
                }
            });
        }
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (eventListener != null) {
            seguindoDataBase.removeEventListener(eventListener);
        }
    }

}

File: app/src/main/java/activity/SeguindoActivity.kt
package activity

import adapter.SeguindoAdapater
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.design.widget.TabLayout
import android.support.v4.view.ViewPager
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.Toast
import br.com.icaro.filme.R
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import domain.API
import domain.TvSeasons
import domain.UserTvshow
import domain.tvshow.Tvshow
import rx.Observer
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import utils.UtilsApp
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Created by icaro on 25/11/16.
 */
class SeguindoActivity : BaseActivity() {

    private var viewPager: ViewPager? = null
    private var tabLayout: TabLayout? = null
    private var progressBar: ProgressBar? = null
    private var linearLayout: LinearLayout? = null

    private var seguindoDataBase: DatabaseReference? = null
    private var atualizarDatabase: FirebaseDatabase? = null
    private var database: FirebaseDatabase? = null
    private var eventListener: ValueEventListener? = null
    private var mAuth: FirebaseAuth? = null

    private var userTvshowFire: MutableList<UserTvshow>? = null
    private var userTvshowNovo: UserTvshow? = null

    private var compositeSubscription: CompositeSubscription? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_usuario_list)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        setUpToolBar()
        supportActionBar?.setTitle(R.string.seguindo)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        viewPager = findViewById<View>(R.id.viewpage_usuario) as ViewPager
        tabLayout = findViewById<View>(R.id.tabLayout) as TabLayout
        progressBar = findViewById<View>(R.id.progress) as ProgressBar
        linearLayout = findViewById<View>(R.id.linear_usuario_list) as LinearLayout


        if (UtilsApp.isNetWorkAvailable(this)) {
            iniciarFirebases()
            setEventListenerSeguindo()
        } else {
            snack()
        }
    }


    private fun iniciarFirebases() {

        mAuth = FirebaseAuth.getInstance()
        database = FirebaseDatabase.getInstance()
        seguindoDataBase = database?.getReference("users")?.child(mAuth?.currentUser!!
                .uid)?.child("seguindo")
    }

    protected fun snack() {
        Snackbar.make(linearLayout!!, R.string.no_internet, Snackbar.LENGTH_INDEFINITE)
                .setAction(R.string.retry) {
                    if (UtilsApp.isNetWorkAvailable(baseContext)) {
                        //text_elenco_no_internet.setVisibility(View.GONE);
                        iniciarFirebases()
                        setEventListenerSeguindo()
                    } else {
                        snack()
                    }
                }.show()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }


    private fun setupViewPagerTabs() {

        viewPager?.offscreenPageLimit = 1
        viewPager?.currentItem = 0
        tabLayout?.setupWithViewPager(viewPager)
        tabLayout?.setSelectedTabIndicatorColor(resources.getColor(R.color.accent))
        viewPager?.adapter = SeguindoAdapater(this@SeguindoActivity, supportFragmentManager,
                userTvshowFire)
    }

    private fun setEventListenerSeguindo() {
        eventListener = object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                userTvshowFire = ArrayList()
                if (dataSnapshot.exists()) {
                    dataSnapshot.children
                            .asSequence()
                            .map { it.getValue(UserTvshow::class.java) }
                            .forEach { userTvshowFire?.add(it!!) }
                    veriricarSerie()
                }
                setupViewPagerTabs()
                progressBar?.visibility = View.GONE
            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        }
        seguindoDataBase?.addListenerForSingleValueEvent(eventListener)
    }

    fun veriricarSerie() {

        userTvshowFire?.forEachIndexed { indexSerie, userTvshow ->
            var serie: Tvshow? = null
            val subscriber = API(this)
                    .getTvShow(userTvshow.id)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(object : Observer<Tvshow> {
                        override fun onCompleted() {
                            if (serie?.numberOfEpisodes != userTvshow.numberOfEpisodes) {
                                userTvshowNovo = UtilsApp.setUserTvShow(serie)
                                userTvshow.seasons.forEachIndexed { indexSeason, userSeasons ->
                                    if (serie?.seasons!![indexSeason]?.episodeCount != userSeasons.userEps.size) {
                                        atualizarRealDate(indexSerie, indexSeason, serie, userTvshow)}
                                }
                            }
                        }

                        override fun onError(e: Throwable) {
                            // Toast.makeText(this@SeguindoActivity, R.string.ops, Toast.LENGTH_SHORT).show()
                            Log.d("TAG", e.message)
                        }

                        override fun onNext(tvshow: Tvshow) {
                            serie = tvshow
                        }
                    })
            compositeSubscription?.add(subscriber)
        }
    }

    fun atualizarRealDate(indexSerie: Int, indexSeason: Int, serie: Tvshow?, userTvshow: UserTvshow) {

        var tvseasonRetorno: TvSeasons? = null
        val subscriber = API(this)
                .getTvSeasons(userTvshow.id, userTvshow.numberOfSeasons)
                .debounce(5000, TimeUnit.MILLISECONDS)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.immediate())
                .subscribe(object : Observer<TvSeasons> {
                    override fun onCompleted() {
                        userTvshowNovo?.seasons?.get(indexSeason)?.userEps = UtilsApp.setEp2(tvseasonRetorno)
                        atulizarDataBase(indexSerie, indexSeason)
                    }

                    override fun onError(e: Throwable) {
                        //Toast.makeText(this@SeguindoActivity, R.string.ops, Toast.LENGTH_SHORT).show()
                        Log.d("TAG", e.message)
                    }

                    override fun onNext(tvseason: TvSeasons) {
                        tvseasonRetorno = tvseason
                    }
                })

        compositeSubscription?.add(subscriber)
    }

    private fun atulizarDataBase(indexSerie: Int, indexSeason: Int) {

        userTvshowNovo?.seasons?.get(indexSeason)?.seasonNumber = userTvshowFire!![indexSerie].seasons[indexSeason].seasonNumber
        userTvshowNovo?.seasons?.get(indexSeason)?.isVisto = userTvshowFire!![indexSerie].seasons[indexSeason].isVisto
        atulizarDataBaseEps(indexSerie, indexSeason)

    }

    private fun atulizarDataBaseEps(indexSerie: Int, indexSeason: Int) {

        for ((indexEp, userEp) in userTvshowFire?.get(indexSerie)?.seasons?.get(indexSeason)?.userEps?.withIndex()!!) {
            userTvshowNovo?.seasons?.get(indexSeason)?.userEps?.set(indexEp, userEp)
        }

        if (userTvshowNovo?.seasons?.get(indexSeason)?.userEps?.size!! > userTvshowFire?.get(indexSerie)?.seasons?.get(indexSeason)!!.userEps.size) {
            userTvshowNovo?.seasons?.get(indexSeason)?.isVisto = false
        }

        setDataBase(userTvshowNovo, indexSerie, indexSeason)
    }

    private fun setDataBase(userTvshowNovo: UserTvshow?, indexSerie: Int, index: Int) {

        val childUpdates = HashMap<String, Any>().apply {
            put("/numberOfEpisodes", userTvshowNovo?.numberOfEpisodes!!)
            put("/numberOfSeasons", userTvshowNovo?.numberOfSeasons)
            put("/poster", userTvshowNovo.poster)
            put("seasons/${index}", userTvshowNovo.seasons!![index])
        }

        atualizarDatabase = FirebaseDatabase.getInstance()
        val myRef = atualizarDatabase?.getReference("users")
        myRef?.child(mAuth?.currentUser?.uid)
                ?.child("seguindo")
                ?.child(userTvshowNovo?.id.toString())
                ?.updateChildren(childUpdates)
                ?.addOnCompleteListener({ task ->
                    if (task.isComplete) {
                        setupViewPagerTabs()
                        Toast.makeText(this@SeguindoActivity, R.string.season_updated, Toast.LENGTH_SHORT).show();
                    } else {
                        Log.d("TAG", task.exception.toString())
                    }
                })


    }

    override fun onDestroy() {
        super.onDestroy()
        if (eventListener != null) {
            seguindoDataBase?.removeEventListener(eventListener!!)
        }
        compositeSubscription?.unsubscribe()
    }
}
